contracts/extensions/OneTxPayment.sol

```diff
@@ -60,108 +51,74 @@ contract OneTxPayment is ColonyExtension {
-    );
-    // Check caller is able to call {add,finalize}Payment and moveFundsBetweenPots on the colony
-    validateCallerPermissions(_callerPermissionDomainId, _callerChildSkillIndex, _domainId);
-    // In addition, check the caller is able to call moveFundsBetweenPots from the root domain
-    require(
-      ColonyAuthority(colony.authority()).canCall(msg.sender, 1, address(colony), MOVE_FUNDS_SIG),
-      "colony-one-tx-payment-root-funding-not-authorized"
-    );
-    // Get around stack too deep with array
-    // paymentData[0] = expenditure slot
-    // paymentData[1] = expenditureId/paymentId
-    // paymentData[3] = fundingPotId
-    uint256[3] memory paymentData;
-    paymentData[0] = 0;
-    if (_tokens.length > 1) {
-      // Make a new expenditure
-      paymentData[1] = colony.makeExpenditure(_permissionDomainId, _childSkillIndex, _domainId);
-      ColonyDataTypes.Expenditure memory expenditure = colony.getExpenditure(paymentData[1]);
-      paymentData[2] = expenditure.fundingPotId;
-      colony.setExpenditureRecipient(paymentData[1], paymentData[0], _workers[0]);
-      colony.setExpenditureSkill(paymentData[1], paymentData[0], _skillId);
+    require(_workers.length == _tokens.length && _workers.length == _amounts.length, "colony-one-tx-payment-invalid-input");
+    validatePermissions(1, _childSkillIndex, _domainId);
+
+    if (_workers.length == 1) {
+
+      uint256 paymentId = colony.addPayment(1, _childSkillIndex, _workers[0], _tokens[0], _amounts[0], _domainId, _skillId);
+      uint256 fundingPotId = colony.getPayment(paymentId).fundingPotId;
+
+      colony.moveFundsBetweenPots(1, UINT256_MAX, _childSkillIndex, 1, fundingPotId, _amounts[0], _tokens[0]);
+
+      colony.finalizePayment(1, _childSkillIndex, paymentId);
+      colony.claimPayment(paymentId, _tokens[0]);
+
     } else {
-      // Add a new payment
-      paymentData[1] = colony.addPayment(
-        _permissionDomainId,
-        _childSkillIndex,
-        _workers[0],
-        _tokens[0],
-        _amounts[0],
-        _domainId,
-        _skillId);
-      ColonyDataTypes.Payment memory payment = colony.getPayment(paymentData[1]);
-      paymentData[2] = payment.fundingPotId;
-    }
 
-    for (uint256 index = 0; index < _workers.length; index++) {
-      if (_tokens.length > 1) {
-        if (index != 0 && _workers[index] != _workers[index-1]) {
-          paymentData[0]++;
-          colony.setExpenditureRecipient(paymentData[1], paymentData[0], _workers[index]);
-          colony.setExpenditureSkill(paymentData[1], paymentData[0], _skillId);
+      uint256 expenditureId = colony.makeExpenditure(1, _childSkillIndex, _domainId);
+      uint256 fundingPotId = colony.getExpenditure(expenditureId).fundingPotId;
+
+      uint256 idx;
+      uint256 slot;
+
+      for (idx = 0; idx < _workers.length; idx++) {
+        colony.moveFundsBetweenPots(1, UINT256_MAX, _childSkillIndex, 1, fundingPotId, _amounts[idx], _tokens[idx]);
+
```

Comment:
Probably also want to do similar with the token addresses for each user to avoid unexpected behaviour - specifying the same user and token address twice does not do what you think it does currently.

Reply:
I was thinking through this, and honestly I think it's best to leave it like this. Making clients sort addresses into a canonical order seems tedious and arbitrary, especially as addresses are randomly generated. I don't see a problem with allowing users to receive multiple skills reputation with one expenditure, since that behavior could be equivalently achieved at higher cost via multiple expenditures. If a caller of this function has the same user/token pair twice in a row, then the previous token value would be overwritten, yes, but I don't see a great way to avoid that. The obvious approach would involve querying for the expenditure, looking at any existing payout, and setting the payout to that sum, but the additional gas overhead that would impose seems very not worth it for preventing this edge case. I'd rather have documentation explaining these possible behaviors and telling users to be careful with how they input their arguments.

Reply:
> I don't see a problem with allowing users to receive multiple skills reputation with one expenditure, since that behavior could be equivalently achieved at higher cost via multiple expenditures. 

I agree, and said as much. I'm saying that if a second address isn't involved, it's not possible to do in single expenditure, and that it is silly that it's possible in one instance and not the other.

> Making clients sort addresses into a canonical order seems tedious and arbitrary 

I do not agree with this. This is exactly the sort of computationally complex to do, but trivial to check is correct, computation that should be offloaded to the clientside in environments that involve on-chain contracts.

Reply:
Alright, inasmuch as it makes sense to make behavior either consistently available or not at all -- achieving behavior, even if not problematic, in a conditional way seems... quirky. I'll assert address order, but I'd rather not do anything about repeated tokens.

Reply:
Why not? Is that behaviour not also quirky?

Reply:
Ok, I was able to figure out a way to introduce this check in a relatively simple way.